File: src/affix-rules.ts
export interface AffixEntry {
  strip: string;
  add: string;
  condition: string;
}

export interface AffixRule {
  type: "PFX" | "SFX";
  flag: string;
  crossProduct: boolean;
  entries: AffixEntry[];
}

export class AffixRules {
  private rules: Map<string, AffixRule[]>;
  private cache: Map<string, Set<string>>;
  private readonly CACHE_SIZE = 10000;

  constructor() {
    this.rules = new Map();
    this.cache = new Map();
  }

  addRule(rule: AffixRule): void {
    const existing = this.rules.get(rule.flag) || [];
    existing.push(rule);
    this.rules.set(rule.flag, existing);
  }

  private getCachedVariations(word: string): Set<string> | undefined {
    return this.cache.get(word);
  }

  private cacheVariations(word: string, variations: Set<string>): void {
    if (this.cache.size >= this.CACHE_SIZE) {
      // Remove oldest entries when cache is full
      const entriesToRemove = Math.floor(this.CACHE_SIZE * 0.1);
      const keys = Array.from(this.cache.keys()).slice(0, entriesToRemove);
      for (const key of keys) {
        this.cache.delete(key);
      }
    }
    this.cache.set(word, variations);
  }

  findBaseWords(word: string): Set<string> {
    // Check cache first
    const cached = this.getCachedVariations(word);
    if (cached) return cached;

    const baseWords = new Set<string>();

    // Try removing prefixes
    for (const [, rules] of this.rules) {
      for (const rule of rules) {
        if (rule.type === "PFX") {
          for (const entry of rule.entries) {
            if (
              word.startsWith(entry.add) &&
              this.matchesCondition(word, entry.condition)
            ) {
              const base = entry.strip + word.slice(entry.add.length);
              baseWords.add(base);
            }
          }
        }
      }
    }

    // Try removing suffixes
    for (const [, rules] of this.rules) {
      for (const rule of rules) {
        if (rule.type === "SFX") {
          for (const entry of rule.entries) {
            if (
              word.endsWith(entry.add) &&
              this.matchesCondition(word, entry.condition)
            ) {
              const base = word.slice(0, -entry.add.length) + entry.strip;
              baseWords.add(base);
            }
          }
        }
      }
    }

    // Cache results
    this.cacheVariations(word, baseWords);

    return baseWords;
  }

  private matchesCondition(word: string, condition: string): boolean {
    if (!condition) return true;
    try {
      return new RegExp(condition).test(word);
    } catch {
      return false;
    }
  }

  generateVariations(word: string): Set<string> {
    const variations = new Set<string>();
    variations.add(word);

    // Apply prefixes
    for (const [, rules] of this.rules) {
      for (const rule of rules) {
        if (rule.type === "PFX") {
          for (const entry of rule.entries) {
            if (this.matchesCondition(word, entry.condition)) {
              if (entry.strip) {
                if (word.startsWith(entry.strip)) {
                  variations.add(entry.add + word.slice(entry.strip.length));
                }
              } else {
                variations.add(entry.add + word);
              }
            }
          }
        }
      }
    }

    // Apply suffixes
    for (const [, rules] of this.rules) {
      for (const rule of rules) {
        if (rule.type === "SFX") {
          for (const entry of rule.entries) {
            if (this.matchesCondition(word, entry.condition)) {
              if (entry.strip) {
                if (word.endsWith(entry.strip)) {
                  variations.add(
                    word.slice(0, -entry.strip.length) + entry.add
                  );
                }
              } else {
                variations.add(word + entry.add);
              }
            }
          }
        }
      }
    }

    return variations;
  }
}

File: rust/src/affix_rules.rs
use rustc_hash::{FxHashMap, FxHashSet};
use smallvec::{smallvec, SmallVec};

#[derive(Clone)]
pub struct AffixEntry {
    pub strip: String,
    pub add: String,
    pub condition: String,
}

#[derive(Clone)]
pub struct AffixRule {
    pub rule_type: String, // "PFX" or "SFX"
    pub flag: String,
    pub cross_product: bool,
    pub entries: Vec<AffixEntry>,
}

pub struct AffixRules {
    rules: FxHashMap<String, SmallVec<[AffixRule; 4]>>,
}

impl Default for AffixRules {
    fn default() -> Self {
        Self::new()
    }
}

impl AffixRules {
    pub fn new() -> Self {
        AffixRules {
            rules: FxHashMap::default(),
        }
    }

    pub fn add_rule(&mut self, rule: AffixRule) {
        self.rules
            .entry(rule.flag.clone())
            .or_insert_with(|| smallvec![])
            .push(rule);
    }

    #[inline]
    pub fn find_base_words_into(&self, word: &str, base_words: &mut SmallVec<[String; 4]>) {
        // Try prefix rules
        for rules in self.rules.values() {
            for rule in rules {
                // Access the field to prevent dead code warning, even if you don't use
                let _ = &rule.cross_product;

                match rule.rule_type.as_str() {
                    "PFX" => {
                        for entry in &rule.entries {
                            // Access the field to prevent dead code warning, even if you don't use
                            let _ = &entry.condition;
                            if word.starts_with(&entry.add)
                                && (entry.strip == "0" || word.len() > entry.add.len())
                            {
                                if let Some(base) = word.get(entry.add.len()..) {
                                    base_words.push(base.to_string());
                                }
                            }
                        }
                    }
                    "SFX" => {
                        for entry in &rule.entries {
                            // Access the field to prevent dead code warning, even if you don't use
                            let _ = &entry.condition;
                            if word.ends_with(&entry.add)
                                && (entry.strip == "0" || word.len() > entry.add.len())
                            {
                                if let Some(base) = word.get(..word.len() - entry.add.len()) {
                                    base_words.push(base.to_string());
                                }
                            }
                        }
                    }
                    _ => {}
                }
            }
        }

        // If no base words found, the word might be a base word itself
        if base_words.is_empty() {
            base_words.push(word.to_string());
        }
    }

    // Keep the original method for backward compatibility
    pub fn find_base_words(&self, word: &str) -> FxHashSet<String> {
        let mut base_words = FxHashSet::default();
        let mut temp: SmallVec<[String; 4]> = smallvec![];
        self.find_base_words_into(word, &mut temp);
        base_words.extend(temp);
        base_words
    }
}
